Simplified team forms

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Simplifies team forms
| Type?         | refacto
| Category?     | BO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Part of #16482
| How to test?  | Simplifying Advanced Parameters -> Employee add/edit Employee options and add/edit profile forms. Everything must look the same.

:notebook: **BC Break**
Backwards compatibility break introduced due to extension of TranslationAwareType by EmployeeType and ProfileType. This means if any module extends those types they will get an exception upon upgrading to PS version containing changes in this PR.
<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/21243)
<!-- Reviewable:end -->

JevgenijVisockij commented:
>     * PrestaShop Addons
Fixed the "Password should be at least %num% characters long."

PrestaShop Addons was not translated before, but since it's a name maybe that was intended?

matks commented:
> > ```
> > * PrestaShop Addons
> > ```
> 
> Fixed the "Password should be at least %num% characters long."
> 
> PrestaShop Addons was not translated before, but since it's a name maybe that was intended?

ping @LouiseBonnard @Julievrz to be sure

Names can be translated. For example there is a city in UK called Dover and French people call it Douvres. Also when alphabet change (like arabic alphabet) names change too.

Julievrz commented:
> > > ```
> > > * PrestaShop Addons
> > > ```
> > 
> > 
> > Fixed the "Password should be at least %num% characters long."
> > PrestaShop Addons was not translated before, but since it's a name maybe that was intended?
> 
> ping @LouiseBonnard @Julievrz to be sure
> 
> Names can be translated. For example there is a city in UK called Dover and French people call it Douvres. Also when alphabet change (like arabic alphabet) names change too.

Hi @matks,
After a quick look at the Arabic and Polish projects, it seems that PrestaShop Addons was not translated. 
However, in Hebrew or in Chinese, for example, "Addons" was translated. So,  I agree with you that translators should have the possibility to translate names. What do you think @LouiseBonnard ?

LouiseBonnard commented:
I also agree, 'PrestaShop Addons' or 'Addons marketplace' (depends on how you want to call it in the project) can be found in numerous strings throughout the software so translators are familiar with it. I can even see the _PrestaShop Addons_ string in _Admin.Dashboard.Feature_. :-)

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -109,52 +139,80 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    $this->getNotBlankConstraint(),
+                    $this->getLengthConstraint(EmployeeEmail::MAX_LENGTH),
+                    new Email([
+                        'message' => $this->trans('This field is invalid', 'Admin.Notifications.Error'),
```
```suggestion
                        'message' => $this->trans('This field is invalid.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -109,52 +139,80 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         if ($options['is_restricted_access']) {
-            $builder->add('change_password', ChangePasswordType::class);
+            $builder->add('change_password', ChangePasswordType::class, [
+                'label' => $this->trans('Change password...', 'messages'),
```
I think we should find a domain for this message. Could you show me a screenshot so I can see where it will appear?

JevgenijVisockij replied:
![image](https://user-images.githubusercontent.com/47781757/101279688-97871300-37cc-11eb-99a2-70c9bdb4831a.png)
Hey, here it is.

Julievrz replied:
Thank you! I was not looking at the good PS version, my bad!

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -266,7 +340,7 @@ private function getLengthConstraint($maxLength, $minLength = null)
     {
         return new NotBlank([
-            'message' => $this->trans('This field cannot be empty', [], 'Admin.Notifications.Error'),
+            'message' => $this->trans('This field cannot be empty', 'Admin.Notifications.Error'),
```
```suggestion
            'message' => $this->trans('This field cannot be empty.', 'Admin.Notifications.Error'),
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+  <div class="col-sm">
+    {# "Change password" button #}
+    {{ form_widget(change_password.change_password_button, {
+      label: 'Change password...'|trans({}, 'messages'),
```
same here, we should avoid to a string without a specific domain.

Julievrz replied:
```suggestion
      label: 'Change password...'|trans({}, 'Admin.Actions'),
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+    <div class="card card-body js-change-password-block d-none">
+      {# Current password input #}
+      {{ ps.form_group_row(change_password.old_password, oldPasswordVars, {
+        label: 'Current password'|trans({}, 'messages'),
```
Same here, could you show me a screenshot, please? I cannot find the string in the software...

Julievrz replied:
```suggestion
        label: 'Current password'|trans({}, 'Admin.Advparameters.Feature'),
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+
+      {# New password first input #}
+      <div class="form-group row">
+        {{ ps.label_with_help('New password'|trans({}, 'messages'), passwordHelpMessage, '', change_password.new_password.children.first.vars.id, true) }}
```
Same here please :)

Julievrz replied:
```suggestion
        {{ ps.label_with_help('New password'|trans({}, 'Admin.Advparameters.Feature'), passwordHelpMessage, '', change_password.new_password.children.first.vars.id, true) }}
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+        </div>
+        <div class="col-sm">
+          {{ form_widget(change_password.generate_password_button, {
+            label: 'Generate password'|trans({}, 'messages'),
```
Same here domain to specify

Julievrz replied:
```suggestion
            label: 'Generate password'|trans({}, 'Admin.Actions'),
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+      {# New password second input (confirmation) #}
+      {# Empty help text needed to render the text container for validation feedback messages #}
+      {{ ps.form_group_row(change_password.new_password.children.second, newPasswordSecondVars, {
+        label: 'Confirm password'|trans({}, 'messages'),
```
Same here, I believe there are all on the same page.

Julievrz replied:
```suggestion
        label: 'Confirm password'|trans({}, 'Admin.Advparameters.Feature'),
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+        <span class="strength-low">{{ 'Invalid'|trans({}, 'messages') }}</span>
+        <span class="strength-medium">{{ 'Okay'|trans({}, 'messages') }}</span>
+        <span class="strength-high">{{ 'Good'|trans({}, 'messages') }}</span>
+        <span class="strength-extreme">{{ 'Fabulous'|trans({}, 'messages') }}</span>
```
We have the same issue here. @LouiseBonnard Any idea why there are so many wordings in the "messages" category here? Maybe there is a reason I am missing.

LouiseBonnard replied:
Not at all! As explained [here](https://docs.google.com/presentation/d/18fCfwr4erxwflhq5vT8TbNOCnYOBj5d3OIJOig4Zxso/edit#slide=id.g711cd0b5aa_0_118), all the strings that have not been assigned a domain so far appear in the _messages_ domain. It is not the expected behavior, you're right, and they are inherited from the previous versions so they should get a more accurate domain each time we stumble upon them during wording reviews. ;-)

Julievrz replied:
That's what I thought! :) Thank you Louise!

Julievrz reviewed (commented):
A lot of strings located in "messages". Maybe we should try and find a proper location for them if possible.

Progi1984 commented:
@JevgenijVisockij Always in WIP ?

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -82,24 +83,53 @@ final class EmployeeType extends AbstractType
      * @param string $defaultAvatarUrl
+     * @param bool $isAddonsConnected
+     * @param int $superAdminProfileId
+     * @param \Symfony\Bundle\FrameworkBundle\Routing\Router $router
```
```suggestion
     * @param Router $router
```

Progi1984 reviewed (commented):
A lot of wording feedback

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -109,52 +139,80 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         if ($options['is_restricted_access']) {
-            $builder->add('change_password', ChangePasswordType::class);
+            $builder->add('change_password', ChangePasswordType::class, [
+                'label' => $this->trans('Change password...', 'messages'),
```
```suggestion
                'label' => $this->trans('Change password...', 'Admin.Actions'),
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+
+      {# Password strength feedback messages - used in JS #}
+      <div class="js-password-strength-feedback d-none">
+        <span class="strength-low">{{ 'Invalid'|trans({}, 'messages') }}</span>
```
```suggestion
        <span class="strength-low">{{ 'Invalid'|trans({}, 'Admin.Advparameters.Help') }}</span>
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+      {# Password strength feedback messages - used in JS #}
+      <div class="js-password-strength-feedback d-none">
+        <span class="strength-low">{{ 'Invalid'|trans({}, 'messages') }}</span>
+        <span class="strength-medium">{{ 'Okay'|trans({}, 'messages') }}</span>
```
```suggestion
        <span class="strength-medium">{{ 'Okay'|trans({}, 'Admin.Advparameters.Help') }}</span>
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+      <div class="js-password-strength-feedback d-none">
+        <span class="strength-low">{{ 'Invalid'|trans({}, 'messages') }}</span>
+        <span class="strength-medium">{{ 'Okay'|trans({}, 'messages') }}</span>
+        <span class="strength-high">{{ 'Good'|trans({}, 'messages') }}</span>
```
```suggestion
        <span class="strength-high">{{ 'Good'|trans({}, 'Admin.Advparameters.Help') }}</span>
```

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,97 @@
+        <span class="strength-low">{{ 'Invalid'|trans({}, 'messages') }}</span>
+        <span class="strength-medium">{{ 'Okay'|trans({}, 'messages') }}</span>
+        <span class="strength-high">{{ 'Good'|trans({}, 'messages') }}</span>
+        <span class="strength-extreme">{{ 'Fabulous'|trans({}, 'messages') }}</span>
```
```suggestion
        <span class="strength-extreme">{{ 'Fabulous'|trans({}, 'Admin.Advparameters.Help') }}</span>
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -109,52 +139,80 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('has_enabled_gravatar', SwitchType::class, [
                 'required' => false,
+                'label' => $this->trans('Enable gravatar', 'Admin.Global'),
```
```suggestion
                'label' => $this->trans('Enable gravatar', 'Admin.Advparameters.Feature'),
```

Julievrz commented:
Hi @JevgenijVisockij ! I added wording suggestions to find a domain for strings located in 'messages'. 

Also, I have a question regarding the wording "Fabulous", "Good", etc...when updating the password. I tried it and when I wrote "123456789" I had "Fabulous". 🤔 
Also, I couldn't make the word "invalid" appear, even though I wrote a password a short password (less than 8 characters)...

I am not sure this is the scope of your PR, or if I should create an issue...or maybe there is an explanation.

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php

Added proper domainh for Change password

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Added proper domainh for Change password

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Added proper domain for Current password

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Added proper domain for New password

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Added proper domain for Confirm password

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Added proper domain for Generate password

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Added proper domain for Invalid

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Added proper domain for 'Okay'

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Added proper domain for 'Good'

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Added proper domain for 'Fabulous'

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php

Added proper domain for 'Enable gravatar'

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php

Removed unneeded full class namespace

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php

Added dot to error message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij commented:
> I also agree, 'PrestaShop Addons' or 'Addons marketplace' (depends on how you want to call it in the project) can be found in numerous strings throughout the software so translators are familiar with it. I can even see the _PrestaShop Addons_ string in _Admin.Dashboard.Feature_. :-)

A quick question, should PrestaShop Addons in Edit Employee form
![image](https://user-images.githubusercontent.com/47781757/101649184-4765b600-3a43-11eb-9aa2-a06f8ee22aea.png)
Also be in Admin.Dashboard.Feature domain or should it be in separate (in this case Admin.Advparameters.Feature) domain?

JevgenijVisockij commented:
> Fabulous

It seems that 

> Hi @JevgenijVisockij ! I added wording suggestions to find a domain for strings located in 'messages'.
> 
> Also, I have a question regarding the wording "Fabulous", "Good", etc...when updating the password. I tried it and when I wrote "123456789" I had "Fabulous". thinking
> Also, I couldn't make the word "invalid" appear, even though I wrote a password a short password (less than 8 characters)...
> 
> I am not sure this is the scope of your PR, or if I should create an issue...or maybe there is an explanation. @matks @MatShir

It seems that this is based on jQuery plugin called passy. And last time it was updated (in PrestaShop not the plugin itself) was 2014. So if want to improve this we can update passy or find new plugin but I don't think it's in the scope of this PR. And I am guessing 123456789 is Fablous because at that point any 9 digit password was good enough.

Julievrz commented:
> > I also agree, 'PrestaShop Addons' or 'Addons marketplace' (depends on how you want to call it in the project) can be found in numerous strings throughout the software so translators are familiar with it. I can even see the _PrestaShop Addons_ string in _Admin.Dashboard.Feature_. :-)
> 
> A quick question, should PrestaShop Addons in Edit Employee form
> ![image](https://user-images.githubusercontent.com/47781757/101649184-4765b600-3a43-11eb-9aa2-a06f8ee22aea.png)
> Also be in Admin.Dashboard.Feature domain or should it be in separate (in this case Admin.Advparameters.Feature) domain?

It should be in Admin.Advparameters.Feature in that case 👍

Julievrz commented:
> > Fabulous
> 
> It seems that
> 
> > Hi @JevgenijVisockij ! I added wording suggestions to find a domain for strings located in 'messages'.
> > Also, I have a question regarding the wording "Fabulous", "Good", etc...when updating the password. I tried it and when I wrote "123456789" I had "Fabulous". thinking
> > Also, I couldn't make the word "invalid" appear, even though I wrote a password a short password (less than 8 characters)...
> > I am not sure this is the scope of your PR, or if I should create an issue...or maybe there is an explanation. @matks @MatShir
> 
> It seems that this is based on jQuery plugin called passy. And last time it was updated (in PrestaShop not the plugin itself) was 2014. So if want to improve this we can update passy or find new plugin but I don't think it's in the scope of this PR. And I am guessing 123456789 is Fablous because at that point any 9 digit password was good enough.

Thank you for your answer @JevgenijVisockij I will create a dedicated issue!

JevgenijVisockij committed:
Simplified team forms

JevgenijVisockij committed:
Moved default page to where it should be

atomiix commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -109,63 +139,89 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'help' => $this->trans(
+                    'Password should be at least %num% characters long.',
+                    'Admin.Advparameters.Help',
+                    ['%num%' => self::PASSWORD_MIN_CHARACTER_AMOUNT]
```
What about reusing `Password::MIN_LENGTH` here?

JevgenijVisockij replied:
Makes sense, added it.

atomiix commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/form.html.twig:
```diff
@@ -59,100 +52,20 @@
-
+        {{ form_row(employeeForm.has_enabled_gravatar) }}
+        {{ form_row(employeeForm.email) }}
         {% set passwordHelpMessage = 'Password should be at least %num% characters long.'|trans({ '%num%': 8 }, 'Admin.Advparameters.Help') %}
```
```suggestion
        {% set passwordMinLength = constant('PrestaShop\\PrestaShop\\Core\\Domain\\Employee\\ValueObject\\Password::MIN_LENGTH') %}
        {% set passwordHelpMessage = 'Password should be at least %num% characters long.'|trans({ '%num%': passwordMinLength }, 'Admin.Advparameters.Help') %}
```

JevgenijVisockij replied:
Thanks! Implemented both changes.

Progi1984 reviewed (commented):
Agreed with feedback from @atomiix

JevgenijVisockij committed:
changed PASSWORD_MIN_CHARACTER_AMOUNT to already exisiting Password::MIN_LENGTH

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/form.html.twig

Using constant instead of raw value

Co-authored-by: atomiix <tbaccelli@gmail.com>

khouloudbelguith commented:
Hi @JevgenijVisockij,

The automatic tests are NOK with this PR.
I tried this `Run TEST_PATH="functional/BO/14_advancedParameters/06_team/employees/01_filterAndQuickEditEmployees.js" URL_FO=http://prestashop/PrestaShop/ npm run specific-test`
But I have errors:
![image](https://user-images.githubusercontent.com/16067358/105161348-922e1f80-5b11-11eb-8870-2b375c9c18fe.png)
**Without PR** => OK

Ping @PrestaShop/qa-automation what do you think?

Thank you!

boubkerbribri commented:
Thanks @khouloudbelguith 
The Selector was improved, so now its fixed on the tests. 

@JevgenijVisockij Thanks for the improvement.

khouloudbelguith commented:
Hi @JevgenijVisockij,

I have some issues:

**First issue**
Steps to reproduce the issue:

0. Disable debug mode
1. Go to BO > Advanced Parameters > Team > Employees Tab > Add new employee page 
2. Set all required fields
3. In the Avatar field > choose any not supported field (.zip, .SQL, .ico)
4. Save 
5. See error: Unclear Alert 
![image](https://user-images.githubusercontent.com/16067358/105352472-38e7ee00-5bee-11eb-890e-e2c292d0bf96.png)
6. Go to BO > Advanced Parameters > Team > Profiles Tab > Add new Profiles page
8. Set all required fields
9. In the Avatar field > Add any not supported file
10. Save
11. See error: Unclear alert 
![image](https://user-images.githubusercontent.com/16067358/105352687-85332e00-5bee-11eb-9877-791a75f2b34e.png)

**Second issue**

Steps to reproduce the issue:
0. Enable Debug mode
1. Go to BO > Advanced Parameters > Team > Employees Tab > Add new employee page 
2. Set all required fields
3. Try to change the Permission profile
4. See an error in the console 
![image](https://user-images.githubusercontent.com/16067358/105354553-01c70c00-5bf1-11eb-9549-2ad36254ed75.png)

In fact, if you try to change the Permission profile to `Translator`, the Default page should be automatically changed to `Products`
Also, change the Permission profile to `Translator`, the Default page should display only pages available.
**Without PR**
![image](https://user-images.githubusercontent.com/16067358/105354804-5b2f3b00-5bf1-11eb-9ba0-a24f49360f6c.png)
![image](https://user-images.githubusercontent.com/16067358/105354844-700bce80-5bf1-11eb-93e4-cc3157a6daa1.png)

**With PR**
![image](https://user-images.githubusercontent.com/16067358/105354881-7dc15400-5bf1-11eb-95d7-0310196f1e47.png)
![image](https://user-images.githubusercontent.com/16067358/105354905-86198f00-5bf1-11eb-8940-cc36974750e0.png)

**Third issue**
1. Go to BO > Advanced Parameters > Team > Employees page
2. Edit the employee created by default (id_employee=1)
3. Check that the **Default page** filed doesn't exist
**Without PR**
![image](https://user-images.githubusercontent.com/16067358/105355445-751d4d80-5bf2-11eb-9c13-7b5a8c274513.png)

**With PR**
![image](https://user-images.githubusercontent.com/16067358/105355480-823a3c80-5bf2-11eb-9c1f-e01542f32112.png)

**Fourth issue**
1. Go to BO > Advanced Parameters > Team > Employees page
2. Edit the employee created by default (id_employee=1)
3. Click on Change password
4. Set current password
5. Set New password with invalid data (only 3 characters) 
![image](https://user-images.githubusercontent.com/16067358/105356319-8a46ac00-5bf3-11eb-9cec-5ff66cb39d04.png)
6. Set Confirm password
7. Save
8. Nothing happens: no alerts displayed

Also, if you set a Confirm password different from the New password & Save => Nothing happens.

Thank you!

JevgenijVisockij committed:
Fixed problem with exception displayed in profile and employee controllers when adding wrong image. Displaying proper message instead.

atomiix commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php:
```diff
@@ -486,6 +489,16 @@ protected function getErrorMessages(Exception $e)
             ),
+            UploadedImageConstraintException::class => [
+                UploadedImageConstraintException::EXCEEDED_SIZE => $this->trans(
+                    'Max file size allowed is "%s" bytes.', 'Admin.Notifications.Error', [
+                    $iniConfig->getUploadMaxSizeInBytes(),
+                ]),
```
```suggestion
                    'Max file size allowed is "%s" bytes.', 'Admin.Notifications.Error',
                    [
                        $iniConfig->getUploadMaxSizeInBytes(),
                    ]
                ),
```

atomiix commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfileController.php:
```diff
@@ -275,7 +276,19 @@ public function bulkDeleteAction(Request $request)
         return [
+            UploadedImageConstraintException::class => [
+                UploadedImageConstraintException::EXCEEDED_SIZE => $this->trans(
+                    'Max file size allowed is "%s" bytes.', 'Admin.Notifications.Error', [
+                    $iniConfig->getUploadMaxSizeInBytes(),
+                ]),
```
```suggestion
                    'Max file size allowed is "%s" bytes.', 'Admin.Notifications.Error',
                    [
                        $iniConfig->getUploadMaxSizeInBytes(),
                    ]
                ),
```

matks replied:
What about 
```suggestion
                    'Max file size allowed is "%s" bytes.', 'Admin.Notifications.Error',
                    [$iniConfig->getUploadMaxSizeInBytes()]
                ),
```

This way we don't add 2 extra lines for brackets while there is a single element inside the array? 🤔  Wdyt ?

JevgenijVisockij replied:
IMO makes sense, thanks you both for suggestion!

JevgenijVisockij committed:
Small position fix

JevgenijVisockij committed:
Fixed issue with exception in console

JevgenijVisockij committed:
Fixed issues with default page not existing

JevgenijVisockij commented:
At the moment it's WIP. There was still one bug I have not fixed, but at the moment githubs "Load more" is not working so I don't know what it is "_".

JevgenijVisockij committed:
Moved multistore feature active back into if(restricted access)

JevgenijVisockij committed:
Restored types to their previous positions

JevgenijVisockij commented:
> Hi @JevgenijVisockij,
> 
> I have some issues:
> 
> **First issue**
> Steps to reproduce the issue:
> 
>     1. Disable debug mode
> 
>     2. Go to BO > Advanced Parameters > Team > Employees Tab > Add new employee page
> 
>     3. Set all required fields
> 
>     4. In the Avatar field > choose any not supported field (.zip, .SQL, .ico)
> 
>     5. Save
> 
>     6. See error: Unclear Alert
>        ![image](https://user-images.githubusercontent.com/16067358/105352472-38e7ee00-5bee-11eb-890e-e2c292d0bf96.png)
> 
>     7. Go to BO > Advanced Parameters > Team > Profiles Tab > Add new Profiles page
> 
>     8. Set all required fields
> 
>     9. In the Avatar field > Add any not supported file
> 
>     10. Save
> 
>     11. See error: Unclear alert
>         ![image](https://user-images.githubusercontent.com/16067358/105352687-85332e00-5bee-11eb-9877-791a75f2b34e.png)
> 
> 
> **Second issue**
> 
> Steps to reproduce the issue:
> 0. Enable Debug mode
> 
>     1. Go to BO > Advanced Parameters > Team > Employees Tab > Add new employee page
> 
>     2. Set all required fields
> 
>     3. Try to change the Permission profile
> 
>     4. See an error in the console
>        ![image](https://user-images.githubusercontent.com/16067358/105354553-01c70c00-5bf1-11eb-9549-2ad36254ed75.png)
> 
> 
> In fact, if you try to change the Permission profile to `Translator`, the Default page should be automatically changed to `Products`
> Also, change the Permission profile to `Translator`, the Default page should display only pages available.
> **Without PR**
> ![image](https://user-images.githubusercontent.com/16067358/105354804-5b2f3b00-5bf1-11eb-9ba0-a24f49360f6c.png)
> ![image](https://user-images.githubusercontent.com/16067358/105354844-700bce80-5bf1-11eb-93e4-cc3157a6daa1.png)
> 
> **With PR**
> ![image](https://user-images.githubusercontent.com/16067358/105354881-7dc15400-5bf1-11eb-95d7-0310196f1e47.png)
> ![image](https://user-images.githubusercontent.com/16067358/105354905-86198f00-5bf1-11eb-8940-cc36974750e0.png)
> 
> **Third issue**
> 
>     1. Go to BO > Advanced Parameters > Team > Employees page
> 
>     2. Edit the employee created by default (id_employee=1)
> 
>     3. Check that the **Default page** filed doesn't exist
>        **Without PR**
>        ![image](https://user-images.githubusercontent.com/16067358/105355445-751d4d80-5bf2-11eb-9c13-7b5a8c274513.png)
> 
> 
> **With PR**
> ![image](https://user-images.githubusercontent.com/16067358/105355480-823a3c80-5bf2-11eb-9c1f-e01542f32112.png)
> 
> **Fourth issue**
> 
>     1. Go to BO > Advanced Parameters > Team > Employees page
> 
>     2. Edit the employee created by default (id_employee=1)
> 
>     3. Click on Change password
> 
>     4. Set current password
> 
>     5. Set New password with invalid data (only 3 characters)
>        ![image](https://user-images.githubusercontent.com/16067358/105356319-8a46ac00-5bf3-11eb-9cec-5ff66cb39d04.png)
> 
>     6. Set Confirm password
> 
>     7. Save
> 
>     8. Nothing happens: no alerts displayed
> 
> 
> Also, if you set a Confirm password different from the New password & Save => Nothing happens.
> 
> **Fifth issue**
> 
>     1. Enable Multistore context
> 
>     2. Create a new shop
> 
>     3. Try to create a new employee and select only one shop
> 
>     4. Save
> 
>     5. Edit the last employee created => see the error
> 
>     6. => the employee cannot be associated with one shop only
> 
> 
> Thank you!

Hey, thanks for Q&A. Solved the first 4 issues.

Regarding fifth couldn't reproduce it. Could you specify what the error message was? I assumed it was " the employee cannot be associated with one shop only" but couldn't find such message anywhere.

JevgenijVisockij committed:
Moved default page where it should be according to structure

JevgenijVisockij committed:
Removed uneeded line

Progi1984 commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,100 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ *#}
+
+{% set passwordHelpMessage = 'Password should be at least %num% characters long.'|trans({ '%num%': 8 }, 'Admin.Advparameters.Help') %}
```
```suggestion
{% set passwordHelpMessage = 'Password should be at least %num% characters long.'|trans({ '%num%': 8 }, 'Admin.Advparameters.Help') %}
```

For the `8` please use a constant.

Progi1984 replied:
Better contribution by @atomiix

Progi1984 reviewed (commented):
A small feedback

khouloudbelguith commented:
Hi @JevgenijVisockij,

>Regarding fifth couldn't reproduce it. Could you specify what the error message was? I assumed it was " the employee cannot be associated with one shop only" but couldn't find such message anywhere.

I don't know exactly the behavior in this case.
The checkboxes in the Advanced Parameters > Team > Employees page exist only in the develop branch (with or without PR).
In the PS1.7.7 => the checkboxes do not exist.

Thank you!

matks commented:
> Hi @JevgenijVisockij,
> 
> > Regarding fifth couldn't reproduce it. Could you specify what the error message was? I assumed it was " the employee cannot be associated with one shop only" but couldn't find such message anywhere.
> 
> I don't know exactly the behavior in this case.
> The checkboxes in the Advanced Parameters > Team > Employees page exist only in the develop branch (with or without PR).
> In the PS1.7.7 => the checkboxes do not exist.
> 
> Thank you!

ping @PrestaShop/product-team @Julievrz 

1. do you know if PrestaShop is supposed to allow, in multishop usecase, to create an employee that is active only for 1 shop ?
That means this employee
- can only login into this shop
- can only see/use BO pages on this shop

2. if this is not possible (according to our tests that should be the usecase), if someone attempts to do it (create an employee that is active only for 1 shop), can you tell us what should be the error message?

Julievrz commented:
Hi @matks, I cannot answer your first question (maybe @marionf has an idea? ☺️) but if you need a wording, here is my suggestion:

'An employee's status cannot be modified for one store only.' in Admin.Parameters.Notification

marionf commented:
@matks 

> do you know if PrestaShop is supposed to allow, in multishop usecase, to create an employee that is active only for 1 shop ?

Yes, this is possible for the default profiles (Logistician, Commercial, Translator) and the new profiles created, as you can see in the screenshot below, you can choose the shop(s) with which the profile will be associated through the shop association block.

![capture d'écran_2943](https://user-images.githubusercontent.com/13449658/106628695-5c118680-657a-11eb-9652-23ab7a5b5d42.png)

However this is not possible for the superadmin profile, as there is no shop association block you can't choose the shop(s) with which it is associated, it's automatically attached to all stores.
Fot this case, the message has already been defined in [the specs](https://github.com/PrestaShop/prestashop-specs/blob/master/broader-topics/multistorespecialsspecs.md):
Display an information message: "Note that this feature is available in all shops context only. It will be added to all your stores."

By the way, I just updated the multistore specs with this page that wasn't in:
https://github.com/PrestaShop/prestashop-specs/pull/204

Julievrz commented:
> @matks @Julievrz
> 
> > do you know if PrestaShop is supposed to allow, in multishop usecase, to create an employee that is active only for 1 shop ?
> 
> Yes, this is possible for the default profiles (Logistician, Commercial, Translator) and the new profiles created, as you can see in the screenshot below, you can choose the shop(s) with which the profile will be associated through the shop association block.
> 
> ![capture d'écran_2943](https://user-images.githubusercontent.com/13449658/106628695-5c118680-657a-11eb-9652-23ab7a5b5d42.png)
> 
> However this is not possible for the superadmin profile, as there is no shop association block you can't choose the shop(s) with which it is associated, it's automatically attached to all stores.
> Fot this case, the message has already been defined in [the specs](https://github.com/PrestaShop/prestashop-specs/blob/master/broader-topics/multistorespecialsspecs.md):
> Display an information message: "Note that this feature is available in all shops context only. It will be added to all your stores."
> 
> By the way, I just updated the multistore specs with this page that wasn't in:
> [PrestaShop/prestashop-specs#204](https://github.com/PrestaShop/prestashop-specs/pull/204)

Does this wording ('Note that this feature is available in all shops context only. It will be added to all your stores." need to be added before I change the label to Wording ✔️?

matks commented:
Hi @khouloudbelguith @marionf @Julievrz @JevgenijVisockij I spent some time to summarize what's happening on this PR 😄 .

## 1. QA feedback

@khouloudbelguith reported the following issue:

- Enable Multistore context
- Create a new shop
- Create a new employee, associate him with only 1 shop
- Save
- Edit the last employee created => see an error

@JevgenijVisockij and I fail to reproduce this behavior 🤔  @khouloudbelguith can you retry? or maybe @Robin-Fischer-PS ? maybe there is something we missed in the setup? a video record could help

## 2. About employees add/edit and multishop

There was some confusion about what can be done or not for employees and multishops. So I summarize it here:

A. It should be impossible to create/edit a SuperAdmin only associated with 1 shop. It always has access to all shops.

➡️  I have checked, it seems to be OK. Only missing item, as mentioned by @marionf, is the warning message "that this feature is available in all shops only". It needs to be added below the "associate with shops" widget.

B. It is possible to create/edit an employee that is not a SuperAdmin to associate him/her only with 1 shop.

➡️  I have checked, it seems to be OK. 

## 3. Next steps

We need a QA analyst to retest this PR and confirm whether @khouloudbelguith feedback is indeed a bug we need to fix.

I think we can seize the opportunity of this PR to add the missing warning message mentioned by @marionf .

marionf commented:
@matks Thanks for the summary :+1: 

> A. It should be impossible to create/edit a SuperAdmin only associated with 1 shop. It always has access to all shops.
I have checked, it seems to be OK. Only missing item, as mentioned by @marionf, is the warning message "that this feature is available in all shops only". It needs to be added below the "associate with shops" widget.

I am not sure what you are calling  the "associate with shops" widget ?

As when you select superadmin as profile, the shop association block isn't displayed, I think the information message (not warning :wink:) "Note that this feature is available in all shops context only. It will be added to all your stores." should rather be displayed at the top of the page :)

matks commented:
> I am not sure what you are calling the "associate with shops" widget ?

The shops association block 😉 
>
> As when you select superadmin as profile, the shop association block isn't displayed, I think the information message (not warning 😉) "Note that this feature is available in all shops context only. It will be added to all your stores." should rather be displayed at the top of the page :)

We'll put the message there, at the place you want, with the color you want 😄

JevgenijVisockij commented:
> Note that this feature is available in all shops context only. It will be added to all your stores.

Hey so I have implemented the feature, just to confirm it's all correct, when you select superadmin it should look like this, and message dissapears when you select any other profile.
![image](https://user-images.githubusercontent.com/47781757/110654984-51748d80-81c7-11eb-974d-8aec563d0d53.png)

Also when I am editing my own profile shop selection is not available, neither is the alert seen. Should I leave it that way?

JevgenijVisockij committed:
Changed wording domain

JevgenijVisockij commented:
> > @matks @Julievrz
> > > do you know if PrestaShop is supposed to allow, in multishop usecase, to create an employee that is active only for 1 shop ?
> > 
> > 
> > Yes, this is possible for the default profiles (Logistician, Commercial, Translator) and the new profiles created, as you can see in the screenshot below, you can choose the shop(s) with which the profile will be associated through the shop association block.
> > ![capture d'écran_2943](https://user-images.githubusercontent.com/13449658/106628695-5c118680-657a-11eb-9652-23ab7a5b5d42.png)
> > However this is not possible for the superadmin profile, as there is no shop association block you can't choose the shop(s) with which it is associated, it's automatically attached to all stores.
> > Fot this case, the message has already been defined in [the specs](https://github.com/PrestaShop/prestashop-specs/blob/master/broader-topics/multistorespecialsspecs.md):
> > Display an information message: "Note that this feature is available in all shops context only. It will be added to all your stores."
> > By the way, I just updated the multistore specs with this page that wasn't in:
> > [PrestaShop/prestashop-specs#204](https://github.com/PrestaShop/prestashop-specs/pull/204)
> 
> Does this wording ('Note that this feature is available in all shops context only. It will be added to all your stores.') need to be added before I change the label to Wording heavy_check_mark?

@Julievrz I have added the help text so you can check it.

marionf commented:
> Hey @marionf  so I have implemented the feature, just to confirm it's all correct, when you select superadmin it should look like this, and message dissapears when you select any other profile.
![image](https://user-images.githubusercontent.com/47781757/110654984-51748d80-81c7-11eb-974d-8aec563d0d53.png)

Hey @JevgenijVisockij, yes that's perfect, thank you :+1: 

> Also when I am editing my own profile shop selection is not available, neither is the alert seen. Should I leave it that way?

It seems that while editing a profile (Logistician, Commercial, Translator or the new profiles created), the shop association block is displayed only if you are connected with a superadmin profile. That's make sense, because otherwise I will be able to edit my own profile and give me access to stores that I am not authorized to access.
So, in this case, I suggest to display the following information message: "Note that this profile will be edited for all the stores with which it is associated", to be checked by @Julievrz

Julievrz commented:
Those wordings suggestions look good to me. 👍

JevgenijVisockij committed:
Added  message when employee edits his own profile

JevgenijVisockij commented:
Added the message

JevgenijVisockij committed:
added radix to parseInt

atomiix commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,100 @@
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ *#}
+
+{% set passwordHelpMessage = 'Password should be at least %num% characters long.'|trans({ '%num%': 8 }, 'Admin.Advparameters.Help') %}
```
```suggestion

{% set passwordMinLength = constant('PrestaShop\\PrestaShop\\Core\\Domain\\Employee\\ValueObject\\Password::MIN_LENGTH') %}
{% set passwordHelpMessage = 'Password should be at least %num% characters long.'|trans({ '%num%': passwordMinLength }, 'Admin.Advparameters.Help') %}
```

JevgenijVisockij replied:
This is nice, didn't know you could add constant like this. Thanks!

atomiix replied:
I just copied what has been done here https://github.com/PrestaShop/PrestaShop/pull/22045/files#diff-dbc3e4500f8b7833c72212c5fa8c5e8c40e1edf2a5a168bdfb722474d3514e81R71 😄

Progi1984 reviewed (commented):
A last one feedback by @atomiix

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig

Using constant from value object in twig

Co-authored-by: atomiix <tbaccelli@gmail.com>

PierreRambaud commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,101 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ *#}
+
+{% set passwordMinLength = constant('PrestaShop\\PrestaShop\\Core\\Domain\\Employee\\ValueObject\\Password::MIN_LENGTH') %}
```
Should be something we can pass as template parameter?

JevgenijVisockij replied:
Good point! moved it there.

khouloudbelguith commented:
Hi @JevgenijVisockij, @matks,

This [Part](https://github.com/PrestaShop/PrestaShop/pull/22045#issuecomment-779902556) is ok :heavy_check_mark: 
The profiles page is ok :heavy_check_mark: 
The employees page is NOK, I found some issues:

**First issue**

1. Go to BO > Advanced Parameters > Team > Employees page
2. Add/Edit an employee => in the `Default page` section: the whole style is changed

**With PR** 
![image](https://user-images.githubusercontent.com/16067358/113138022-8d7a9b80-9225-11eb-8f0f-2b5e510c3cef.png)
**Without PR**
![image](https://user-images.githubusercontent.com/16067358/113138177-bbf87680-9225-11eb-802c-088c005e1de5.png)

**Second issue**
1. Go to BO > Advanced Parameters > Team > Employees page
2. Edit the employee by default
3. Click on `Change password`
4. Set current password
5. Set New password with `0123456789`
6. See error => you have `invalid` as an alert => NOK 
![image](https://user-images.githubusercontent.com/16067358/113138440-11cd1e80-9226-11eb-8c87-30915c321289.png)
7. Set Confirm password with `0123456789`
8. Click Save => ok

=> the alert in the step 6 is NOK

**Third issue**

In the `Employee options`, the data cannot be saved
1. If you try to set `Password regeneration` to 5 and save
2. if you reload page => not saved
3. If you try to enable `Memorize the language used in Admin panel forms` and save
4. If you reload the page => it is still disabled

**Fourth issue**

The automatic tests are NOK
I run this command: `TEST_PATH="functional/BO/14_advancedParameters/06_team/employees/01_filterAndQuickEditEmployees.js" URL_FO=http://prestashop/PrestaShop/ npm run specific-test`
=> NOK
[employees.pdf](https://github.com/PrestaShop/PrestaShop/files/6236274/employees.pdf)


Thank you!

JevgenijVisockij committed:
Fixed page search not working

JevgenijVisockij committed:
Moved password min length setting from template to controller

JevgenijVisockij committed:
Added info how to run unit tests

JevgenijVisockij committed:
Changed password to weak instead of invalid

JevgenijVisockij commented on admin-dev/themes/new-theme/js/components/change-password-handler.js:
```diff
@@ -103,7 +103,7 @@ export default class ChangePasswordHandler {
           message: this.$feedbackContainer.find('.strength-low').text(),
-          elementClass: 'text-danger',
+          elementClass: 'text-warning',
         };
```
Changing this because I am also changing low streingth warning from "invalid" to "weak". 

I think that makes more sense because even if password is "invalid" according to passy(the library responsible for password strength), it will still be saved. 
In general version of passy in PrestaShop is quite outdated and hardly makes sense ("123456789" is a "fabulous" password, but 0123456789 is "invalid") and needs to be changed/removed/updated but I think it's out of scope for this PR.

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php:
```diff
@@ -486,6 +492,16 @@ protected function getErrorMessages(Exception $e)
             ),
+            UploadedImageConstraintException::class => [
+                UploadedImageConstraintException::EXCEEDED_SIZE => $this->trans(
+                    'Max file size allowed is "%s" bytes.', 'Admin.Notifications.Error',
```
```suggestion
                    'Max file size allowed is "%d" bytes.', 'Admin.Notifications.Error',
```
@JevgenijVisockij @Progi1984 So is it supposed to be "%d" here?

Progi1984 replied:
Yes

JevgenijVisockij commented:
> Hi @JevgenijVisockij, @matks,
> 
> This [Part](https://github.com/PrestaShop/PrestaShop/pull/22045#issuecomment-779902556) is ok heavy_check_mark
> The profiles page is ok heavy_check_mark
> The employees page is NOK, I found some issues:
> 
> **First issue**
> 
>     1. Go to BO > Advanced Parameters > Team > Employees page
> 
>     2. Add/Edit an employee => in the `Default page` section: the whole style is changed
> 
> 
> **With PR**
> ![image](https://user-images.githubusercontent.com/16067358/113138022-8d7a9b80-9225-11eb-8f0f-2b5e510c3cef.png)
> **Without PR**
> ![image](https://user-images.githubusercontent.com/16067358/113138177-bbf87680-9225-11eb-802c-088c005e1de5.png)
> 
> **Second issue**
> 
>     1. Go to BO > Advanced Parameters > Team > Employees page
> 
>     2. Edit the employee by default
> 
>     3. Click on `Change password`
> 
>     4. Set current password
> 
>     5. Set New password with `0123456789`
> 
>     6. See error => you have `invalid` as an alert => NOK
>        ![image](https://user-images.githubusercontent.com/16067358/113138440-11cd1e80-9226-11eb-8c87-30915c321289.png)
> 
>     7. Set Confirm password with `0123456789`
> 
>     8. Click Save => ok
> 
> 
> => the alert in the step 6 is NOK
> 
> **Third issue**
> 
> In the `Employee options`, the data cannot be saved
> 
>     1. If you try to set `Password regeneration` to 5 and save
> 
>     2. if you reload page => not saved
> 
>     3. If you try to enable `Memorize the language used in Admin panel forms` and save
> 
>     4. If you reload the page => it is still disabled
> 
> 
> **Fourth issue**
> 
> The automatic tests are NOK
> I run this command: `TEST_PATH="functional/BO/14_advancedParameters/06_team/employees/01_filterAndQuickEditEmployees.js" URL_FO=http://prestashop/PrestaShop/ npm run specific-test`
> => NOK
> [employees.pdf](https://github.com/PrestaShop/PrestaShop/files/6236274/employees.pdf)
> Ping @PrestaShop/qa-automation
> 
> Thank you!

Fixed first 3 issues.(regarding password I changed it to from "Invalid" to "weak" so it doesn't incorrectly indicate that you can't save form.  Not sure about automated tests when tested didn't see any issues with filters and editing customers.

Progi1984 commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig:
```diff
@@ -89,9 +89,20 @@
+{% block form_row_attr -%}
+  {% set classes = 'form-group row' %}
+  {% if (not compound or force_error|default(false)) and not valid %}
+  {% set classes = classes ~ ' has-error' %}
```
```suggestion
    {% set classes = classes ~ ' has-error' %}
```

Progi1984 reviewed (commented):
Small feedbacks

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig

Added a space for nicer formatting

Co-authored-by: Progi1984 <progi1984@gmail.com>

Progi1984 commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig:
```diff
@@ -68,7 +68,7 @@
     {% endif %}
 
-    <div class="{{ block('form_row_class') }}{% if (not compound or force_error|default(false)) and not valid %} has-error{% endif %}"{% if (attr.visible is defined and not attr.visible) %} style="display: none;"{% endif %}>
+    <div {{ block('form_row_attr') }}{% if (attr.visible is defined and not attr.visible) %} style="display: none;"{% endif %} >
```
```suggestion
    <div {{ block('form_row_attr') }}{% if (attr.visible is defined and not attr.visible) %} style="display: none;"{% endif %}>
```

Progi1984 reviewed (commented):
Small feedbacks

Julievrz commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/form.html.twig:
```diff
@@ -31,6 +32,20 @@
+  {% else %}
+    <div id="multistore-alert" class="alert alert-info" role="alert">
+      <div class="alert-text">
+        {{ 'Note that this feature is available in all shops context only. It will be added to all your stores.'|trans({}, 'Admin.Advparameters.Help') }}<br>
```
```suggestion
        {{ 'Note that this feature is available in all shops context only. It will be added to all your stores.'|trans({}, 'Admin.Notification.Info') }}<br>
```
If this is used elsewhere in the software (and I have noticed on other PRs that it is) it should be in a more generic domain. :-)

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/form.html.twig

Changed translation domain for string

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php

Changed to %d as it's supposed to be only decimal in this string

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfileController.php:
```diff
@@ -275,7 +276,19 @@ public function bulkDeleteAction(Request $request)
         return [
+            UploadedImageConstraintException::class => [
+                UploadedImageConstraintException::EXCEEDED_SIZE => $this->trans(
+                    'Max file size allowed is "%s" bytes.', 'Admin.Notifications.Error',
```
```suggestion
                    'Max file size allowed is "%d" bytes.', 'Admin.Notifications.Error',
```

JevgenijVisockij committed:
Update src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfileController.php

Changed from string to decimal

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

atomiix commented on admin-dev/themes/new-theme/js/app/utils/init-components.js:
```diff
@@ -37,6 +37,7 @@ import TextWithLengthCounter from '@components/form/text-with-length-counter';
 import PreviewOpener from '@components/form/preview-opener';
 import MultistoreConfigField from '@js/components/form/multistore-config-field.js';
 import {EventEmitter} from '@components/event-emitter';
+import ChangePasswordControl from '../../components/form/change-password-control';
```
```suggestion
import ChangePasswordControl from '@components/form/change-password-control';
```

Progi1984 commented:
@JevgenijVisockij Please rebase your PR

JevgenijVisockij committed:
Added TextWithRecommendedLengthCounter component

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php

Added dot to the message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

PierreRambaud commented on admin-dev/themes/new-theme/js/app/utils/init-components.js:
```diff
@@ -38,6 +38,7 @@ import PreviewOpener from '@components/form/preview-opener';
 import MultistoreConfigField from '@js/components/form/multistore-config-field.js';
 import {EventEmitter} from '@components/event-emitter';
 import Router from '@components/router';
+import ChangePasswordControl from '../../components/form/change-password-control';
```
```suggestion
import ChangePasswordControl from '@js/components/form/change-password-control';
```

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/employee/EmployeeForm.js:
```diff
@@ -23,9 +23,7 @@
  */
 
-import ChoiceTree from '../../components/form/choice-tree';
 import AddonsConnector from '../../components/addons-connector';
```
```suggestion
import AddonsConnector from '@components/addons-connector';
```

PierreRambaud commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php:
```diff
@@ -486,6 +492,16 @@ protected function getErrorMessages(Exception $e)
+                    [$iniConfig->getUploadMaxSizeInBytes()]
+                ),
+                UploadedImageConstraintException::UNRECOGNIZED_FORMAT => $this->trans(
+                    'Image format not recognized, allowed formats are: .gif, .jpg, .png',
```
```suggestion
                    'Image format not recognized, allowed formats are: .%s',
```

Should it be this? formats have to came from a const? What if we authorized .jpeg, or .svg or .bmp?

Julievrz replied:
Thank you @PierreRambaud, I agree with you + it will be only 1 string to translate for translators.

JevgenijVisockij replied:
Hmm I simply reused string that has beend widely used in other places, but okay I will refactor so available formats are a const instead.

PierreRambaud commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfileController.php:
```diff
@@ -285,7 +286,19 @@ public function bulkDeleteAction(Request $request)
+                    [$iniConfig->getUploadMaxSizeInBytes()]
+                ),
+                UploadedImageConstraintException::UNRECOGNIZED_FORMAT => $this->trans(
+                    'Image format not recognized, allowed formats are: .gif, .jpg, .png',
```
Same as previous about extensions

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -109,63 +137,89 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'label' => $this->trans('Avatar', 'Admin.Global'),
                 'required' => false,
                 'attr' => [
                     'accept' => 'gif,jpg,jpeg,jpe,png',
```
Should be a const, It's weird because the error message hasn't the same accepted formats :thinking:

JevgenijVisockij replied:
I have made it a const and made sure message containts same formats as here. Right now I added constants to each type individually, I guess same image types are used in many places so it would make sense for one class to constain const's for all the usages, but I am not sure which one. ImageValidator maybe?

Progi1984 commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig:
```diff
@@ -79,7 +79,7 @@
     {% endif %}
 
-    <div class="{{ block('form_row_class') }}{{ block('widget_type_class') }}{% if (not compound or force_error|default(false)) and not valid %} has-error{% endif %}{% if (attr.visible is defined and not attr.visible) %} d-none{% endif %}">
+    <div {{ block('form_row_attr') }} {% if (attr.visible is defined and not attr.visible) %} style="display: none;"{% endif %}>
```
```suggestion
    <div {{ block('form_row_attr') }}{% if (attr.visible is defined and not attr.visible) %} style="display: none;"{% endif %}>
```

Progi1984 reviewed (commented):
Small feedbacks

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig

Removed uneeded space

Co-authored-by: Progi1984 <progi1984@gmail.com>

Julievrz commented:
@JevgenijVisockij @matks There's still a wording change to make following @PierreRambaud comment. 
https://github.com/PrestaShop/PrestaShop/pull/22045#discussion_r652419553

Progi1984 commented:
@JevgenijVisockij Don't forget to rebase.

JevgenijVisockij committed:
Made shops block dissapear when superadmin selected

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig

# Conflicts:
#	admin-dev/themes/new-theme/js/pages/employee/EmployeeForm.ts

JevgenijVisockij committed:
Added information message to display when editing superadmin

# Conflicts:
#	admin-dev/themes/new-theme/js/pages/employee/EmployeeForm.ts

JevgenijVisockij committed:
Added password control component to init components

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.js

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.js

JevgenijVisockij committed:
Moved router in correct import order

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.js

JevgenijVisockij committed:
Update admin-dev/themes/new-theme/js/app/utils/init-components.js

Changed component inclusion to use @js

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Removed duplicate password control

JevgenijVisockij committed:
Changed error messeges to use constants

JevgenijVisockij committed:
Fixed type script not working

JevgenijVisockij committed:
Cs fixer fixes

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -109,63 +147,89 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('has_enabled_gravatar', SwitchType::class, [
                 'required' => false,
+                'label' => $this->trans('Enable gravatar', 'Admin.Advparameters.Feature'),
```
```suggestion
                'label' => $this->trans('Enable gravatar', 'Admin.Global'),
```

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php

Updated domain

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

Progi1984 commented on templates/bundles/PrestaShopBundle/Admin/Configure/AdvancedParameters/Profiles/Blocks/form.html.twig:
```diff
@@ -33,16 +33,7 @@
-      }) }}
-
+      {{ form_widget(profileForm) }}
       {% if avatarUrl|default('') is not empty %}
```
But with behavior the avatar is at the last and not anymore after the name ?

Progi1984 reviewed (commented):
@JevgenijVisockij Thank you for your contribution. I have added a comment, and I saw the CI is red for ESLint errors.

PierreRambaud commented:
Hi @JevgenijVisockij 
Should we close it since no one working on it? And reopen if needed?
The code is outdated :(

JevgenijVisockij committed:
Added behat test for positions

JevgenijVisockij committed:
Fixed translation

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php

JevgenijVisockij committed:
Removed no longer needed connect to addons

Progi1984 commented:
@JevgenijVisockij :pray: Needs a rebase

sowbiba commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -166,23 +212,43 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'label' => $this->trans('Default page', 'Admin.Advparameters.Feature'),
+                'help' => $this->trans('This page will be displayed just after login.', 'Admin.Advparameters.Help'),
+                'attr' => [
+                    'data-minimumResultsForSearch' => '7',
```
candidate for a constant ?

sowbiba commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/form.html.twig:
```diff
@@ -55,128 +62,16 @@
-        {% set passwordHelpMessage = 'Password should be at least %num% characters long.'|trans({ '%num%': 8 }, 'Admin.Advparameters.Help') %}
+        {{ form_row(employeeForm.has_enabled_gravatar) }}
+        {{ form_row(employeeForm.email) }}
+        {% set passwordHelpMessage = 'Password should be at least %num% characters long.'|trans({ '%num%': passwordMinLength }, 'Admin.Advparameters.Help') %}
```
I don't see where this variable is used

Progi1984 replied:
```suggestion
```

sowbiba reviewed (commented):
1 suggestion and 1 question ;)

sowbiba committed:
ES-Lint fix

sowbiba committed:
YAML Lint fixes

sowbiba committed:
Fix path to prestashop_ui_kit template

matks reviewed (commented):
Nothing to add, when @sowbiba feedbacks are applied I approve 👍

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -166,23 +212,43 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'label' => $this->trans('Default page', 'Admin.Advparameters.Feature'),
+                'help' => $this->trans('This page will be displayed just after login.', 'Admin.Advparameters.Help'),
+                'attr' => [
+                    'data-minimumResultsForSearch' => '7',
```
```suggestion
                    'data-minimum-results-for-search' => '7',
```
But yes, it requires a const :thinking:

sowbiba replied:
If I search for minimumResultsForSearch, I have a lot results with camelCase. I'm wondering if it's not because jquery.select requires it like that

Progi1984 commented on admin-dev/themes/new-theme/js/pages/employee/EmployeeForm.ts:
```diff
@@ -23,9 +23,8 @@
 
-import ChoiceTree from '../../components/form/choice-tree';
+import ChoiceTree from '@components/form/choice-tree';
 import AddonsConnector from '../../components/addons-connector';
```
```suggestion
import AddonsConnector from '@components/addons-connector';
```

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -44,16 +44,25 @@
+        'jpg',
+        'jpeg',
+        'jpe',
+        'png',
```
```suggestion
        'png',
        'webp',
```

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -44,16 +44,25 @@
+        'png',
+    ];
+
+    public const AVAILABLE_IMAGE_FORMATS_STRING_FOR_TRANSLATION = 'gif, jpg, jpeg, jpe, png';
```
```suggestion
    public const AVAILABLE_IMAGE_FORMATS_STRING_FOR_TRANSLATION = 'gif, jpg, jpeg, jpe, png, webp';
```

Progi1984 commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig:
```diff
@@ -79,7 +79,7 @@
     {% endif %}
 
-    <div class="{{ block('form_row_class') }}{{ block('widget_type_class') }}{% if (not compound or force_error|default(false)) and not valid %} has-error{% endif %}{% if (attr.visible is defined and not attr.visible) %} d-none{% endif %}">
+    <div {{ block('form_row_attr') }}{% if (attr.visible is defined and not attr.visible) %} style="display: none;"{% endif %}>
```
Why using inline-css and not class `d-none` ?

Progi1984 reviewed (commented):
@JevgenijVisockij Thank you for your contribution. I have brought some feedbacks.

sowbiba committed:
Add webp image type, use d-none instead of inline css

sowbiba committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/form.html.twig

sowbiba committed:
Remove unneeded constant

Progi1984 commented:
@JevgenijVisockij Could you rebase your PR & remove merge commits, please ?

jolelievre commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -44,16 +44,26 @@
  * Class EmployeeType defines an employee form.
  */
-final class EmployeeType extends AbstractType
+final class EmployeeType extends TranslatorAwareType
```
```suggestion
class EmployeeType extends TranslatorAwareType
```

jolelievre replied:
We avoid final now to help module extend our code

jolelievre commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -108,53 +139,69 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    $this->getNotBlankConstraint(),
+                    $this->getLengthConstraint(EmployeeEmail::MAX_LENGTH),
+                    new Email([
+                        'message' => $this->trans('This field is invalid.', 'Admin.Notifications.Error'),
```
Not sure this is needed since `Email` constraint from Symfony already has its own message which should be translated by the framework
But nevermind it's fine this way

jolelievre commented on src/PrestaShopBundle/Form/Admin/Configure/AdvancedParameters/Employee/EmployeeType.php:
```diff
@@ -108,53 +139,69 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         ;
 
         if ($options['is_restricted_access']) {
+            $builder->add('change_password', ChangePasswordType::class, [
```
The field is added two times with the same field name, so this one is useless, it's gonna be overwritten right after and all these options will be ignored

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/change_password.html.twig:
```diff
@@ -0,0 +1,100 @@
+
+    <div class="card card-body js-change-password-block d-none">
+      {# Current password input #}
+      {{ ps.form_group_row(change_password.old_password, oldPasswordVars, {
```
This doesn't simplify the form since every fields is added one by one manually This should be the expected result:

jolelievre replied:
```suggestion
      {{ form_row(change_password) }}
```

jolelievre replied:
If the `ChangedPasswordType` neds a specific layout it should be handled in the UI kit theme

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/employee_options.html.twig:
```diff
@@ -34,16 +36,12 @@
         <div class="form-group row">
           <label for="{{ employeeOptionsForm.password_change_time.vars.id }}" class="form-control-label">
-            {{ 'Password regeneration'|trans({}, 'Admin.Advparameters.Feature') }}
+            {{ form_label(employeeOptionsForm.password_change_time) }}
```
Same here the form doesn't seem to be simplified

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Employee/Blocks/form.html.twig:
```diff
@@ -36,17 +51,9 @@
-        {{ ps.form_group_row(employeeForm.avatarUrl, {}, {
-          'label': 'Avatar'|trans({}, 'Admin.Global')
-        }) }}
+        {{ form_row(employeeForm.firstname) }}
```
Same here it should only be:

jolelievre replied:
```suggestion
        {{ form_row(employeeForm) }}
```

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/Profiles/Blocks/form.html.twig:
```diff
@@ -33,16 +33,7 @@
-        'label': 'Avatar'|trans({}, 'Admin.Global')
-      }) }}
-
+      {{ form_widget(profileForm) }}
```
Niceone! That's the expected result for other forms as well

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig:
```diff
@@ -110,6 +110,32 @@
-{% block form_row_class -%}
-  form-group row{% if row_attr.class is defined %} {{ row_attr.class }}{% endif %}
-{%- endblock form_row_class %}
+{% block form_row_attr -%}
```
This is a good idea, however, you should keep the `form_row_class` block and include it because some modules extend it It's an easy way to extend our UI kit and change all the rows' classes at once

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig:
```diff
@@ -110,6 +110,32 @@
-  form-group row{% if row_attr.class is defined %} {{ row_attr.class }}{% endif %}
-{%- endblock form_row_class %}
+{% block form_row_attr -%}
+  {% set classes = 'form-group row' %}
```
```suggestion
  {% set classes = block('form_row_class')|trim %}
```

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig:
```diff
@@ -110,6 +110,32 @@
+
+  {% set row_attr = row_attr|merge({'class': classes}) %}
+
+  {% for key, rowAttr in row_attr %} {{ key }}="{{ rowAttr }}" {% endfor %}
```
That's a super feature, I also realized recently that `row_attr` could only be used for the `class` option
Now we're talking!

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig:
```diff
@@ -110,6 +110,32 @@
+  {% set row_attr = row_attr|merge({'class': classes}) %}
+
+  {% for key, rowAttr in row_attr %} {{ key }}="{{ rowAttr }}" {% endfor %}
+{%- endblock form_row_attr %}
```
However, this whole block should be put in the `prestashop_ui_kit_base.html.twig` form theme so that this new attribute block can also be used in the other layout (which is used for product page)

The only diff will be that the base layout doesn't use the same class for its row it uses `form-group` instead of `form-group row` This is where keeping a separated block `form_row_class` will come in handy, the base layout can define it And this layout can simply override it to include its different classes The `form_row_attr` will remain common and the same for both themes

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig:
```diff
@@ -110,6 +110,32 @@
+  {% for key, rowAttr in row_attr %} {{ key }}="{{ rowAttr }}" {% endfor %}
+{%- endblock form_row_attr %}
+
+{%- block repeated_row -%}
```
I'm not sure about the purpose of this block?

jolelievre reviewed (commented):
Thanks @JevgenijVisockij 

A few comments here and there, I love the form row attributes block!! This will be useful everywhere, but it needs to be moved into the base layout

Julievrz reviewed (approved):
Wording ✔️

Progi1984 commented:
Hi @JevgenijVisockij 
Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.
Thanks!
